File: objective-c-sample/SendBirdSampleUI/SendBird/SendBirdMessaging/View/MessagingChannelTableViewCell.m
//
//  MessagingChannelTableViewCell.m
//  SendBirdSampleUI
//
//  Created by SendBird Developers on 2015. 7. 29..
//  Copyright (c) 2015 SENDBIRD.COM. All rights reserved.
//

#import "MessagingChannelTableViewCell.h"

#define kChannelUrlFontSize 14.0
#define kChannelMembersFontSize 11.0
#define kChannelCoverRadius 19.0
#define kChannelLastMessageFontSize 11.0
#define kChannelLastMessageDateFontSize 9.0
#define kChannelUnreadCountFontSize 11.0

@implementation MessagingChannelTableViewCell {
    NSLayoutConstraint *memberCountImageViewWidthConstraint;
}

- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    
    if (self != nil) {
        [self initViews];
    }
    
    return self;
}

- (void) initViews
{
    [self setSelectionStyle:UITableViewCellSelectionStyleNone];
    
    self.profileImageView = [[UIImageView alloc] init];
    [self.profileImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.profileImageView.layer setCornerRadius:kChannelCoverRadius];
    [self.profileImageView setClipsToBounds:YES];
    [self.profileImageView setContentMode:UIViewContentModeScaleAspectFill];
    [self addSubview:self.profileImageView];
    
    self.nicknameLabel = [[UILabel alloc] init];
    [self.nicknameLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.nicknameLabel setFont:[UIFont boldSystemFontOfSize:kChannelUrlFontSize]];
    [self.nicknameLabel setTextColor:UIColorFromRGB(0x3d3d3d)];
    [self addSubview:self.nicknameLabel];
    
    self.lastMessageLabel = [[UILabel alloc] init];
    [self.lastMessageLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.lastMessageLabel setFont:[UIFont systemFontOfSize:kChannelLastMessageFontSize]];
    [self.lastMessageLabel setTextColor:UIColorFromRGB(0x999999)];
    [self addSubview:self.lastMessageLabel];
    
    self.bottomLineView = [[UIView alloc] init];
    [self.bottomLineView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.bottomLineView setBackgroundColor:UIColorFromRGB(0xc8c8c8)];
    [self addSubview:self.bottomLineView];
    
    self.lastMessageDateLabel = [[UILabel alloc] init];
    [self.lastMessageDateLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.lastMessageDateLabel setFont:[UIFont systemFontOfSize:kChannelLastMessageDateFontSize]];
    [self.lastMessageDateLabel setTextColor:UIColorFromRGB(0x999999)];
    [self.lastMessageDateLabel setText:@"Apr 16, 2015"];
    [self addSubview:self.lastMessageDateLabel];
    
    self.unreadCountImageView = [[UIImageView alloc] init];
    [self.unreadCountImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.unreadCountImageView setImage:[UIImage imageNamed:@"_bg_notify"]];
    [self addSubview:self.unreadCountImageView];
    
    self.unreadCountLabel = [[UILabel alloc] init];
    [self.unreadCountLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.unreadCountLabel setFont:[UIFont systemFontOfSize:kChannelUnreadCountFontSize]];
    [self.unreadCountLabel setTextColor:UIColorFromRGB(0xffffff)];
    [self.unreadCountLabel setText:@"99"];
    [self addSubview:self.unreadCountLabel];
    
    self.checkImageView = [[UIImageView alloc] init];
    [self.checkImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.checkImageView setImage:[UIImage imageNamed:@"_btn_check_off"]];
    [self addSubview:self.checkImageView];
    
    self.memberCountImageView = [[UIImageView alloc] init];
    [self.memberCountImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.memberCountImageView setImage:[UIImage imageNamed:@"_icon_group_number" ]];
    [self addSubview:self.memberCountImageView];
    
    self.memberCountLabel = [[UILabel alloc] init];
    [self.memberCountLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.memberCountLabel setText:@"5"];
    [self.memberCountLabel setTextColor:UIColorFromRGB(0xa4acbc)];
    [self.memberCountLabel setFont:[UIFont systemFontOfSize:10.0]];
    //    [self.memberCountLabel setBackgroundColor:[UIColor redColor]];
    [self addSubview:self.memberCountLabel];
    
    
    // Profile Image View
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.profileImageView
                                                     attribute:NSLayoutAttributeCenterY
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeCenterY
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.profileImageView
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:12]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.profileImageView
                                                     attribute:NSLayoutAttributeWidth
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:40]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.profileImageView
                                                     attribute:NSLayoutAttributeHeight
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:40]];
    
    // Nickname Label
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.nicknameLabel
                                                     attribute:NSLayoutAttributeTop
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTop
                                                    multiplier:1 constant:10]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.nicknameLabel
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.profileImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:14]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.nicknameLabel
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.memberCountImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-10]];
    
    // Last Message Label
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.lastMessageLabel
                                                     attribute:NSLayoutAttributeTop
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.nicknameLabel
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:6]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.lastMessageLabel
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.profileImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:14]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.lastMessageLabel
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-46]];
    
    // Last Message Date
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.lastMessageDateLabel
                                                     attribute:NSLayoutAttributeTop
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTop
                                                    multiplier:1 constant:12]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.lastMessageDateLabel
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-12]];
    
    // Unread Count Background Image View
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.unreadCountImageView
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:-10]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.unreadCountImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-12]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.unreadCountImageView
                                                     attribute:NSLayoutAttributeWidth
                                                     relatedBy:NSLayoutRelationGreaterThanOrEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:22]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.unreadCountImageView
                                                     attribute:NSLayoutAttributeHeight
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:22]];
    
    // Unread Count Label
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.unreadCountLabel
                                                     attribute:NSLayoutAttributeCenterX
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.unreadCountImageView
                                                     attribute:NSLayoutAttributeCenterX
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.unreadCountLabel
                                                     attribute:NSLayoutAttributeCenterY
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.unreadCountImageView
                                                     attribute:NSLayoutAttributeCenterY
                                                    multiplier:1 constant:0]];
    
    // Bottom Line View
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.bottomLineView
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.bottomLineView
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:66]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.bottomLineView
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.bottomLineView
                                                     attribute:NSLayoutAttributeHeight
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:0.5]];
    
    // Check Image View
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.checkImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-12]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.checkImageView
                                                     attribute:NSLayoutAttributeCenterY
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeCenterY
                                                    multiplier:1 constant:0]];
    
    // Member Count View
    // Member Count ImageView
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.memberCountImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-70]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.memberCountImageView
                                                     attribute:NSLayoutAttributeCenterY
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.nicknameLabel
                                                     attribute:NSLayoutAttributeCenterY
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.memberCountImageView
                                                     attribute:NSLayoutAttributeHeight
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:12.5]];
    memberCountImageViewWidthConstraint = [NSLayoutConstraint constraintWithItem:self.memberCountImageView
                                                                       attribute:NSLayoutAttributeWidth
                                                                       relatedBy:NSLayoutRelationGreaterThanOrEqual
                                                                          toItem:nil
                                                                       attribute:NSLayoutAttributeNotAnAttribute
                                                                      multiplier:1 constant:22];
    [self addConstraint:memberCountImageViewWidthConstraint];
    // Member Count Label
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.memberCountLabel
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.memberCountImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-2]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.memberCountLabel
                                                     attribute:NSLayoutAttributeCenterY
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.memberCountImageView
                                                     attribute:NSLayoutAttributeCenterY
                                                    multiplier:1 constant:0]];
}

- (void) setSelected:(BOOL)selected
{
    [super setSelected:selected];
    
    if (selected) {
        [self.checkImageView setImage:[UIImage imageNamed:@"_btn_check_on"]];
    }
    else {
        [self.checkImageView setImage:[UIImage imageNamed:@"_btn_check_off"]];
    }
}

- (void) setModel:(SendBirdMessagingChannel *)model withCheckMark:(BOOL)check
{
    if (model == nil) {
        return;
    }
    
    SendBirdMemberInMessagingChannel *member = nil;
    if ([[model members] count] > 0) {
        member = [[model members] objectAtIndex:0];
    }
    
    NSString *channelName = [NSString stringWithFormat:@"%@", [SendBirdUtils getMessagingChannelNames:[model members]]];
    [self.nicknameLabel setText:channelName];
    [self.lastMessageLabel setText:[[model lastMessage] message]];
    
    if (check) {
        [self.unreadCountLabel setHidden:YES];
        [self.unreadCountImageView setHidden:YES];
        [self.lastMessageDateLabel setHidden:YES];
        [self.checkImageView setHidden:NO];
    }
    else {
        [self.checkImageView setHidden:YES];
        if ([model lastMessage] == nil) {
            [self.lastMessageDateLabel setHidden:YES];
        }
        else {
            [self.lastMessageDateLabel setHidden:NO];
            long long ts = [[model lastMessage] getMessageTimestamp] / 1000;
            [self.lastMessageDateLabel setText:[SendBirdUtils lastMessageDateTime:ts]];
        }
        
        int unreadCount = [model unreadMessageCount];
        if (unreadCount > 0) {
            [self.unreadCountImageView setHidden:NO];
            [self.unreadCountLabel setHidden:NO];
            NSString *unreadCountText = @"";
            if (unreadCount < 99) {
                unreadCountText = [NSString stringWithFormat:@"%d", unreadCount];
            }
            else {
                unreadCountText = @"99";
            }
            [self.unreadCountLabel setText:unreadCountText];
        }
        else {
            [self.unreadCountImageView setHidden:YES];
            [self.unreadCountLabel setHidden:YES];
        }
    }
    
    [SendBirdUtils loadImage:[SendBirdUtils getDisplayCoverImageUrl:[model members]] imageView:self.profileImageView width:40 height:40];

    if ([[model members] count] > 2) {
        [self.memberCountLabel setHidden:NO];
        [self.memberCountImageView setHidden:NO];
        NSString *memberCount = [NSString stringWithFormat:@"%lu", (unsigned long)[[model members] count]];
        
        [self.memberCountLabel setText:memberCount];
        CGRect memberCountRect;
        NSMutableDictionary *memberCountAttribute = [NSMutableDictionary dictionaryWithObjectsAndKeys:[UIFont systemFontOfSize:10.0], NSFontAttributeName, nil];
        NSAttributedString *attributedMemberCount = [[NSAttributedString alloc] initWithString:memberCount attributes:memberCountAttribute];
        memberCountRect = [attributedMemberCount boundingRectWithSize:CGSizeMake(INT64_MAX, 12.5) options:(NSStringDrawingUsesLineFragmentOrigin) context:nil];
        
        memberCountImageViewWidthConstraint.constant = 16 + memberCountRect.size.width;
        [self updateConstraints];
    }
    else {
        [self.memberCountLabel setHidden:YES];
        [self.memberCountImageView setHidden:YES];
    }
}

@end

File: swift-sample/SendBirdSampleUI/SendBird/SendBirdMessaging/View/MessagingChannelTableViewCell.swift
//
//  MessagingChannelTableViewCell.swift
//  SendBirdSampleUI
//
//  Created by Jed Kyung on 2/7/16.
//  Copyright © 2016 SENDBIRD.COM. All rights reserved.
//

import UIKit
import SendBirdSDK

class MessagingChannelTableViewCell: UITableViewCell {
    let kChannelUrlFontSize: CGFloat = 14.0
    let kChannelMembersFontSize: CGFloat = 11.0
    let kChannelCoverRadius: CGFloat = 19.0
    let kChannelLastMessageFontSize: CGFloat = 11.0
    let kChannelLastMessageDateFontSize: CGFloat = 9.0
    let kChannelUnreadCountFontSize: CGFloat = 11.0
    
    var profileImageView: UIImageView?
    var nicknameLabel: UILabel?
    var lastMessageLabel: UILabel?
    var bottomLineView: UIView?
    var unreadCountImageView: UIImageView?
    var unreadCountLabel: UILabel?
    var lastMessageDateLabel: UILabel?
    var checkImageView: UIImageView?
    var memberCountImageView: UIImageView?
    var memberCountLabel: UILabel?
    
    private var memberCountImageViewWidthConstraint: NSLayoutConstraint?

    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        self.initViews()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private func initViews() {
        self.selectionStyle = UITableViewCellSelectionStyle.None
        
        self.profileImageView = UIImageView()
        self.profileImageView?.translatesAutoresizingMaskIntoConstraints = false
        self.profileImageView?.layer.cornerRadius = kChannelCoverRadius
        self.profileImageView?.clipsToBounds = true
        self.profileImageView?.contentMode = UIViewContentMode.ScaleAspectFill
        self.addSubview(self.profileImageView!)
        
        self.nicknameLabel = UILabel()
        self.nicknameLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.nicknameLabel?.font = UIFont.boldSystemFontOfSize(kChannelUrlFontSize)
        self.nicknameLabel?.textColor = SendBirdUtils.UIColorFromRGB(0x3d3d3d)
        self.addSubview(self.nicknameLabel!)
        
        self.lastMessageLabel = UILabel()
        self.lastMessageLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.lastMessageLabel?.font = UIFont.systemFontOfSize(kChannelLastMessageFontSize)
        self.lastMessageLabel?.textColor = SendBirdUtils.UIColorFromRGB(0x999999)
        self.addSubview(self.lastMessageLabel!)
        
        self.bottomLineView = UIView()
        self.bottomLineView?.translatesAutoresizingMaskIntoConstraints = false
        self.bottomLineView?.backgroundColor = SendBirdUtils.UIColorFromRGB(0xc8c8c8)
        self.addSubview(self.bottomLineView!)
        
        self.lastMessageDateLabel = UILabel()
        self.lastMessageDateLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.lastMessageDateLabel?.font = UIFont.systemFontOfSize(kChannelLastMessageDateFontSize)
        self.lastMessageDateLabel?.textColor = SendBirdUtils.UIColorFromRGB(0x999999)
        self.lastMessageDateLabel?.text = "Apr 16, 2015"
        self.addSubview(self.lastMessageDateLabel!)
        
        self.unreadCountImageView = UIImageView()
        self.unreadCountImageView?.translatesAutoresizingMaskIntoConstraints = false
        self.unreadCountImageView?.image = UIImage.init(named: "_bg_notify")
        self.addSubview(self.unreadCountImageView!)
        
        self.unreadCountLabel = UILabel()
        self.unreadCountLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.unreadCountLabel?.font = UIFont.systemFontOfSize(kChannelUnreadCountFontSize)
        self.unreadCountLabel?.textColor = SendBirdUtils.UIColorFromRGB(0xffffff)
        self.unreadCountLabel?.text = "99"
        self.addSubview(self.unreadCountLabel!)
        
        self.checkImageView = UIImageView()
        self.checkImageView?.translatesAutoresizingMaskIntoConstraints = false
        self.checkImageView?.image = UIImage.init(named: "_btn_check_off")
        self.addSubview(self.checkImageView!)
        
        self.memberCountImageView = UIImageView()
        self.memberCountImageView?.translatesAutoresizingMaskIntoConstraints = false
        self.memberCountImageView?.image = UIImage.init(named: "_icon_group_number")
        self.addSubview(self.memberCountImageView!)
        
        self.memberCountLabel = UILabel()
        self.memberCountLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.memberCountLabel?.text = "5"
        self.memberCountLabel?.textColor = SendBirdUtils.UIColorFromRGB(0xa4acbc)
        self.memberCountLabel?.font = UIFont.systemFontOfSize(10.0)
        self.addSubview(self.memberCountLabel!)
        
        // Profile Image View
        self.addConstraint(NSLayoutConstraint.init(item: self.profileImageView!, attribute: NSLayoutAttribute.CenterY, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.CenterY, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.profileImageView!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: 12))
        self.addConstraint(NSLayoutConstraint.init(item: self.profileImageView!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 40))
        self.addConstraint(NSLayoutConstraint.init(item: self.profileImageView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 40))
        
        // Nickname Label
        self.addConstraint(NSLayoutConstraint.init(item: self.nicknameLabel!, attribute: NSLayoutAttribute.Top, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Top, multiplier: 1, constant: 10))
        self.addConstraint(NSLayoutConstraint.init(item: self.nicknameLabel!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self.profileImageView!, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: 14))
        self.addConstraint(NSLayoutConstraint.init(item: self.nicknameLabel!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self.memberCountImageView!, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -10))
        
        // Last Message Label
        self.addConstraint(NSLayoutConstraint.init(item: self.lastMessageLabel!, attribute: NSLayoutAttribute.Top, relatedBy: NSLayoutRelation.Equal, toItem: self.nicknameLabel!, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: 6))
        self.addConstraint(NSLayoutConstraint.init(item: self.lastMessageLabel!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self.profileImageView!, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: 14))
        self.addConstraint(NSLayoutConstraint.init(item: self.lastMessageLabel!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -46))
        
        // Last Message Date
        self.addConstraint(NSLayoutConstraint.init(item: self.lastMessageDateLabel!, attribute: NSLayoutAttribute.Top, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Top, multiplier: 1, constant: 12))
        self.addConstraint(NSLayoutConstraint.init(item: self.lastMessageDateLabel!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -12))
        // Unread Count Background Image View
        self.addConstraint(NSLayoutConstraint.init(item: self.unreadCountImageView!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: -10))
        self.addConstraint(NSLayoutConstraint.init(item: self.unreadCountImageView!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -12))
        self.addConstraint(NSLayoutConstraint.init(item: self.unreadCountImageView!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 22))
        self.addConstraint(NSLayoutConstraint.init(item: self.unreadCountImageView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 22))
        
        // Unread Count Label
        self.addConstraint(NSLayoutConstraint.init(item: self.unreadCountLabel!, attribute: NSLayoutAttribute.CenterX, relatedBy: NSLayoutRelation.Equal, toItem: self.unreadCountImageView!, attribute: NSLayoutAttribute.CenterX, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.unreadCountLabel!, attribute: NSLayoutAttribute.CenterY, relatedBy: NSLayoutRelation.Equal, toItem: self.unreadCountImageView!, attribute: NSLayoutAttribute.CenterY, multiplier: 1, constant: 0))
        
        // Bottom Line View
        self.addConstraint(NSLayoutConstraint.init(item: self.bottomLineView!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.bottomLineView!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: 66))
        self.addConstraint(NSLayoutConstraint.init(item: self.bottomLineView!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.bottomLineView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 0.5))
        
        // Check Image View
        self.addConstraint(NSLayoutConstraint.init(item: self.checkImageView!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -12))
        self.addConstraint(NSLayoutConstraint.init(item: self.checkImageView!, attribute: NSLayoutAttribute.CenterY, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.CenterY, multiplier: 1, constant: 0))
        
        // Member Count View
        // Member Count ImageView
        self.addConstraint(NSLayoutConstraint.init(item: self.memberCountImageView!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -70))
        self.addConstraint(NSLayoutConstraint.init(item: self.memberCountImageView!, attribute: NSLayoutAttribute.CenterY, relatedBy: NSLayoutRelation.Equal, toItem: self.nicknameLabel!, attribute: NSLayoutAttribute.CenterY, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.memberCountImageView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 12.5))
        self.memberCountImageViewWidthConstraint = NSLayoutConstraint.init(item: self.memberCountImageView!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 22)
        self.addConstraint(self.memberCountImageViewWidthConstraint!)
        
        // Member Count Label
        self.addConstraint(NSLayoutConstraint.init(item: self.memberCountLabel!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self.memberCountImageView!, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -2))
        self.addConstraint(NSLayoutConstraint.init(item: self.memberCountLabel!, attribute: NSLayoutAttribute.CenterY, relatedBy: NSLayoutRelation.Equal, toItem: self.memberCountImageView!, attribute: NSLayoutAttribute.CenterY, multiplier: 1, constant: 0))
    }
    
    override func setSelected(selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)
        
        if selected {
            self.checkImageView!.image = UIImage.init(named: "_btn_check_on")
        }
        else {
            self.checkImageView!.image = UIImage.init(named: "_btn_check_off")
        }
    }
    
    func setModel(model: SendBirdMessagingChannel?, check: Bool) {
        if model == nil {
            return
        }
        
        let channelName: String = String.init(format: "%@", SendBirdUtils.getMessagingChannelNames((model?.members)! as NSArray as! [SendBirdMemberInMessagingChannel]))
        self.nicknameLabel?.text = channelName
        self.lastMessageLabel?.text = model?.lastMessage.message
        
        if check {
            self.unreadCountLabel?.hidden = true
            self.unreadCountImageView?.hidden = true
            self.lastMessageDateLabel?.hidden = true
            self.checkImageView?.hidden = false
        }
        else {
            self.checkImageView?.hidden = true
            if model?.lastMessage == nil {
                self.lastMessageDateLabel?.hidden = true
            }
            else {
                self.lastMessageDateLabel?.hidden = false
                let ts: Int64 = (model?.lastMessage.getMessageTimestamp())! / 1000
                self.lastMessageDateLabel?.text = SendBirdUtils.lastMessageDateTime(NSTimeInterval(ts))
            }
            
            let unreadCount: Int32 = model!.unreadMessageCount
            if unreadCount > 0 {
                self.unreadCountImageView?.hidden = false
                self.unreadCountLabel?.hidden = false
                var unreadCountText: String = ""
                if unreadCount < 99 {
                    unreadCountText = String.init(format: "%d", unreadCount)
                }
                else {
                    unreadCountText = "99"
                }
                self.unreadCountLabel?.text = unreadCountText
            }
            else {
                self.unreadCountImageView?.hidden = true
                self.unreadCountLabel?.hidden = true
            }
        }

        SendBirdUtils.loadImage(SendBirdUtils.getDisplayCoverImageUrl((model?.members)! as NSArray as! [SendBirdMemberInMessagingChannel]), imageView: self.profileImageView!, width: 40, height: 40)
        
        if model?.members.count > 2 {
            self.memberCountLabel?.hidden = false
            self.memberCountImageView?.hidden = false
            let memberCount: String = String.init(format: "%lu", (model?.members.count)!)
            
            self.memberCountLabel?.text = memberCount
            var memberCountRect: CGRect?
            let memberCountAttribute = [NSFontAttributeName: UIFont.systemFontOfSize(10.0)]
            let attributedMemberCount: NSAttributedString = NSAttributedString.init(string: memberCount, attributes: memberCountAttribute)
            memberCountRect = attributedMemberCount.boundingRectWithSize(CGSizeMake(CGFloat.max, 12.5), options: NSStringDrawingOptions.UsesLineFragmentOrigin, context: nil)
            
            self.memberCountImageViewWidthConstraint?.constant = 16 + memberCountRect!.size.width
            self.updateConstraints()
        }
        else {
            self.memberCountLabel?.hidden = true
            self.memberCountImageView?.hidden = true
        }
    }
}
